File: Modules/RuleEngine/src/hap/ruleengine/parts/property/BaseProperty.java
package hap.ruleengine.parts.property;


import hap.ruleengine.parts.Component;

public abstract class BaseProperty<T> implements IPropertyContainer
{
	protected final Component myComp;
	protected final String myKey;
	protected final T myDefaultValue;
	private final String myHeader;

	BaseProperty( String header, String key, T defaultValue, Component c )
	{
		myHeader = header;
		myKey = key;
		myDefaultValue = defaultValue;
		myComp = c;
	}

	public String getHeader()
	{
		return myHeader;
	}

	public abstract void setValue(T value);
	public abstract T getValue();
}

File: Modules/RuleEngine/src/hap/ruleengine/parts/property/BaseProperty.kt
package hap.ruleengine.parts.property


import hap.ruleengine.parts.Component
import tornadofx.ViewModel
import tornadofx.getProperty
import tornadofx.onChange
import tornadofx.property

abstract class BaseProperty<T> internal constructor(val header: String, protected val key: String, protected val defaultValue: T, protected val comp: Component) : ViewModel(), IPropertyContainer {
    var value: T by property(defaultValue)
    fun valueProperty() = getProperty(BaseProperty<T>::value)

    abstract fun updateValue(v: T)
    abstract fun readValue() : T

    fun init()
    {
        value = readValue()
    }

    init {
        valueProperty().onChange {
            updateValue(value)
        }
    }
}